File: backend/utils/nlp.js


/**
 * Fallback parser for offline/no-key usage to extract clinical entities
 * using simple heuristics.
 */
function mockExtractClinicalEvent(text) {
  const normalized = text.toLowerCase();
  
  // 1. Check for medication changes
  if (normalized.includes('start') || normalized.includes('give') || normalized.includes('administer') || normalized.includes('hold') || normalized.includes('stop')) {
    let status = 'started';
    if (normalized.includes('hold') || normalized.includes('held')) status = 'held';
    if (normalized.includes('stop') || normalized.includes('discontinue')) status = 'stopped';
    
    // Check known drugs from rules
    const drugs = ['meropenem', 'eliquis', 'solu-medrol', 'methylprednisolone', 'pantocid', 'lasix', 'levosalbutamol', 'budesonide', 'heparin', 'linezolid', 'norad', 'piptaz', 'vasopressin'];
    let foundDrug = 'Unknown Medication';
    for (const d of drugs) {
      if (normalized.includes(d)) {
        foundDrug = d.charAt(0).toUpperCase() + d.slice(1);
        break;
      }
    }

    // Try parsing dose, route, freq
    let dose = 'standard';
    let route = 'oral';
    let freq = 'daily';
    
    if (normalized.includes('iv')) route = 'IV';
    if (normalized.includes('sc') || normalized.includes('subcutaneous')) route = 'SC';
    if (normalized.includes('neb')) route = 'NEB';
    
    if (normalized.includes('1g') || normalized.includes('1 g')) dose = '1g';
    if (normalized.includes('500mg') || normalized.includes('500 mg')) dose = '500mg';
    if (normalized.includes('5mg') || normalized.includes('5 mg')) dose = '5mg';
    if (normalized.includes('40mg') || normalized.includes('40 mg')) dose = '40mg';

    if (normalized.includes('tds') || normalized.includes('three times')) freq = 'TDS';
    if (normalized.includes('bd') || normalized.includes('twice')) freq = 'BD';
    if (normalized.includes('od') || normalized.includes('once')) freq = 'OD';
    if (normalized.includes('qds')) freq = 'QDS';

    return {
      eventType: 'MEDICATION',
      extractedEntities: {
        drugName: foundDrug,
        dose,
        route,
        freq,
        status,
        reason: status === 'held' ? 'active bleeding / clinical safety' : 'clinical management'
      },
      confidence: 0.9,
      flags: status === 'held' ? ['medication_hold'] : [],
      summarySnippet: `Medication ${foundDrug} ${status} (${dose} ${route} ${freq})`
    };
  }

  // 2. Check for investigations
  if (normalized.includes('hb') || normalized.includes('hemoglobin') || normalized.includes('creatinine') || normalized.includes('crp') || normalized.includes('wbc') || normalized.includes('platelet')) {
    let labName = 'Unknown Investigation';
    let value = 'normal';
    let unit = '';
    let status = 'Normal';
    let flags = [];

    if (normalized.includes('hb') || normalized.includes('hemoglobin')) {
      labName = 'Hemoglobin (Hb)';
      unit = 'g/dL';
      const match = normalized.match(/(?:hb|hemoglobin)\s*(?:is|of)?\s*([0-9.]+)/);
      if (match) value = match[1];
      else value = '7.5'; // fallback mock value
      if (parseFloat(value) < 9.0) {
        status = 'Critical Low';
        flags.push('critical_value');
      }
    } else if (normalized.includes('creatinine')) {
      labName = 'Creatinine';
      unit = 'mg/dL';
      const match = normalized.match(/creatinine\s*(?:is|of)?\s*([0-9.]+)/);
      if (match) value = match[1];
      else value = '2.4'; // fallback mock value
      if (parseFloat(value) > 1.5) {
        status = 'High';
        flags.push('high_value');
      }
    } else if (normalized.includes('crp')) {
      labName = 'CRP';
      unit = 'mg/L';
      const match = normalized.match(/crp\s*(?:is|of)?\s*([0-9.]+)/);
      if (match) value = match[1];
      else value = '45';
      if (parseFloat(value) > 10.0) status = 'Elevated';
    }

    return {
      eventType: 'INVESTIGATION',
      extractedEntities: {
        labName,
        value,
        unit,
        status,
        notes: `Extracted from results.`
      },
      confidence: 0.95,
      flags,
      summarySnippet: `${labName} level: ${value} ${unit} (${status})`
    };
  }

  // 3. Check for procedures
  if (normalized.includes('intubat') || normalized.includes('extubat') || normalized.includes('cath') || normalized.includes('dialysis') || normalized.includes('bronchoscopy')) {
    let procedureName = 'Procedure';
    if (normalized.includes('intubat')) procedureName = 'Endotracheal Intubation';
    else if (normalized.includes('extubat')) procedureName = 'Extubation';
    else if (normalized.includes('dialysis') || normalized.includes('hemodialysis')) procedureName = 'Hemodialysis';
    else if (normalized.includes('bronchoscopy')) procedureName = 'Flexible Bronchoscopy';
    else if (normalized.includes('cath')) procedureName = 'Central Venous Catheterization';

    return {
      eventType: 'PROCEDURE',
      extractedEntities: {
        procedureName,
        status: 'Completed',
        notes: text
      },
      confidence: 0.9,
      flags: [],
      summarySnippet: `Procedure performed: ${procedureName}`
    };
  }

  // 4. Check for consultations
  if (normalized.includes('consult') || normalized.includes('referral') || normalized.includes('specialist')) {
    let specialist = 'General Consultation';
    if (normalized.includes('nephro')) specialist = 'Nephrology Consult';
    else if (normalized.includes('cardio')) specialist = 'Cardiology Consult';
    else if (normalized.includes('neuro')) specialist = 'Neurology Consult';

    return {
      eventType: 'CONSULTATION',
      extractedEntities: {
        specialty: specialist,
        notes: text
      },
      confidence: 0.85,
      flags: [],
      summarySnippet: `Specialist Consultation: ${specialist}`
    };
  }

  // 5. Check for diagnoses
  if (normalized.includes('diagnos') || normalized.includes('admit') || normalized.includes('present with') || normalized.includes('pneumonia') || normalized.includes('bleeding') || normalized.includes('hemorrhage')) {
    let diagnosisName = 'Clinical Condition';
    if (normalized.includes('pneumonia')) diagnosisName = 'Pneumonia';
    else if (normalized.includes('hemorrhage') || normalized.includes('bleeding')) diagnosisName = 'Diffuse Alveolar Hemorrhage';
    else if (normalized.includes('ckd') || normalized.includes('kidney')) diagnosisName = 'Chronic Kidney Disease';

    return {
      eventType: 'DIAGNOSIS',
      extractedEntities: {
        diagnosisName,
        status: 'Active',
        notes: text
      },
      confidence: 0.85,
      flags: [],
      summarySnippet: `Diagnosis recorded: ${diagnosisName}`
    };
  }

  // Default fallback to note
  return {
    eventType: 'NOTE',
    extractedEntities: {
      content: text
    },
    confidence: 0.8,
    flags: [],
    summarySnippet: `Clinical note logged.`
  };
}

/**
 * Core clinical entity extractor utilizing Gemini API if API key is present.
 * Otherwise, falls back to local rule-based extractor.
 */
export async function extractClinicalEvent(text, customApiKey = null) {
  const apiKey = customApiKey || process.env.GEMINI_API_KEY;
  if (!apiKey) {
    console.log('No Gemini API key found. Using local heuristic extraction...');
    return mockExtractClinicalEvent(text);
  }

  try {
    const response = await fetch(
      `https://generativelanguage.googleapis.com/v1beta/models/gemini-2.5-flash:generateContent?key=${apiKey}`,
      {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          contents: [
            {
              parts: [
                {
                  text: `You are an expert clinical documentation AI. Parse the following clinical text and extract structured medical event data in JSON format.
                  
Your output MUST be a valid JSON object matching this schema:
{
  "eventType": "DIAGNOSIS" | "MEDICATION" | "PROCEDURE" | "INVESTIGATION" | "CONSULTATION" | "NOTE",
  "extractedEntities": {
    // For DIAGNOSIS:
    "diagnosisName": string,
    "status": string ("Active", "Resolved", "Suspected"),
    // For MEDICATION:
    "drugName": string,
    "dose": string,
    "route": string,
    "freq": string,
    "status": string ("started", "held", "stopped", "continued"),
    "reason": string,
    // For PROCEDURE:
    "procedureName": string,
    "status": string,
    // For INVESTIGATION:
    "labName": string,
    "value": string,
    "unit": string,
    "status": string ("Normal", "High", "Low", "Critical Low"),
    // For CONSULTATION:
    "specialty": string,
    // For General NOTE:
    "content": string,
    "notes": string
  },
  "confidence": number (0.0 to 1.0),
  "flags": string[] (e.g. "critical_value", "medication_hold", "conflict"),
  "summarySnippet": string (A one-line clinical summary sentence, e.g. "Eliquis held due to active pulmonary bleeding.")
}

Return ONLY raw JSON. Do not include markdown code block formatting.

Clinical text to parse:
"${text}"`
                }
              ]
            }
          ],
          generationConfig: {
            responseMimeType: "application/json"
          }
        }),
      }
    );

    if (!response.ok) {
      throw new Error(`Gemini API returned status ${response.status}`);
    }

    const data = await response.json();
    const resultText = data.candidates?.[0]?.content?.parts?.[0]?.text;
    
    // Clean JSON markdown markers if any
    let cleanedText = resultText.trim();
    if (cleanedText.startsWith('```json')) {
      cleanedText = cleanedText.substring(7);
    }
    if (cleanedText.endsWith('```')) {
      cleanedText = cleanedText.substring(0, cleanedText.length - 3);
    }
    
    return JSON.parse(cleanedText.trim());
  } catch (error) {
    console.error('Error in Gemini API extraction, falling back to heuristics:', error);
    return mockExtractClinicalEvent(text);
  }
}

/**
 * Generate Narrative Prose from compiled patient timeline and events
 */
export async function generateNarrative(patientDetails, timelineEvents, type = 'courseInHospital', customApiKey = null) {
  const apiKey = customApiKey || process.env.GEMINI_API_KEY;
  
  // Format timeline text for the prompt or fallback
  const timelineDescription = timelineEvents.map(e => {
    const dateStr = new Date(e.timestamp).toLocaleDateString('en-US', { day: 'numeric', month: 'short', year: 'numeric' });
    const data = JSON.parse(e.eventData);
    return `[${dateStr}] (${e.eventType}) by ${e.authorName}: ${e.provenance} - ${e.eventType === 'NOTE' ? data.noteText : JSON.stringify(data)}`;
  }).join('\n');

  if (!apiKey) {
    console.log('No Gemini API key found for narrative. Generating template-based narrative...');
    
    // Simple heuristic narrative generator
    const meds = timelineEvents.filter(e => e.eventType === 'MEDICATION');
    const labs = timelineEvents.filter(e => e.eventType === 'INVESTIGATION');
    const procs = timelineEvents.filter(e => e.eventType === 'PROCEDURE');
    const consults = timelineEvents.filter(e => e.eventType === 'CONSULTATION');

    if (type === 'medicationJourney') {
      let medText = `Active Medications Tracker:\n`;
      meds.forEach((m, idx) => {
        const data = JSON.parse(m.eventData);
        medText += `${idx + 1}. ${data.drugName} ${data.dose} ${data.route || ''} ${data.freq || ''} - ${data.status.toUpperCase()} (${data.reason || 'Management'})\n`;
      });
      return medText || 'No medications recorded yet.';
    }

    if (type === 'investigationJourney') {
      let labText = `Laboratory Trends:\n`;
      labs.forEach(l => {
        const data = JSON.parse(l.eventData);
        labText += `- ${new Date(l.timestamp).toLocaleDateString()}: ${data.labName} ${data.value} ${data.unit || ''} (${data.status})\n`;
      });
      return labText || 'No labs recorded yet.';
    }

    if (type === 'procedureJourney') {
      let procText = `Procedures Ledger:\n`;
      procs.forEach(p => {
        const data = JSON.parse(p.eventData);
        procText += `- ${new Date(p.timestamp).toLocaleDateString()}: ${data.procedureName} - ${data.status}\n`;
      });
      return procText || 'No procedures performed yet.';
    }

    // Default Course in Hospital
    return `Patient ${patientDetails.name} (${patientDetails.age}yo ${patientDetails.gender}), admitted on ${new Date(patientDetails.admissionDate).toLocaleDateString()} with comorbidities: ${patientDetails.comorbidities}. Currently bedded in ${patientDetails.bed || 'Ward'}.\n\n` +
      `Course Summary:\n` +
      timelineEvents.map(e => {
        const data = JSON.parse(e.eventData);
        if (e.eventType === 'DIAGNOSIS') return `Admitted and diagnosed with ${data.diagnosisName} (${data.severity || ''}).`;
        if (e.eventType === 'MEDICATION') return `Medication ${data.drugName} was ${data.status} (${data.dose} ${data.route || ''} ${data.freq || ''}).`;
        if (e.eventType === 'INVESTIGATION') return `Investigation showed ${data.labName} of ${data.value} ${data.unit || ''} (${data.status}).`;
        if (e.eventType === 'PROCEDURE') return `Procedure performed: ${data.procedureName}.`;
        if (e.eventType === 'CONSULTATION') return `Consultation obtained: ${data.specialty}.`;
        return data.noteText || 'Clinical progress updated.';
      }).join('\n');
  }

  try {
    let promptInstruction = '';
    if (type === 'courseInHospital') {
      promptInstruction = `Generate a detailed, continuous, chronological narrative of the patient's hospital course in formal medical prose. Summarize diagnoses, clinical deterioration, stabilization, medication logic, laboratory results, and investigations. Include specialist consultations. Format as a cohesive paragraph.`;
    } else if (type === 'medicationJourney') {
      promptInstruction = `Generate a summarized Medication Journey. List and explain started, held, and stopped medications with their respective clinical reasons based on the timeline.`;
    } else if (type === 'investigationJourney') {
      promptInstruction = `Generate a summary of laboratory and diagnostic investigation trends. Focus on key biometrics (e.g., Creatinine, Hemoglobin, CRP) and mention values chronologically.`;
    } else {
      promptInstruction = `Generate a chronological list of procedures performed on the patient, detailing their status and indications.`;
    }

    const response = await fetch(
      `https://generativelanguage.googleapis.com/v1beta/models/gemini-2.5-flash:generateContent?key=${apiKey}`,
      {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          contents: [
            {
              parts: [
                {
                  text: `You are an elite pulmonology and critical care clinical documentation assistant.
                  
Patient Details:
- Name: ${patientDetails.name}
- Age: ${patientDetails.age}
- Gender: ${patientDetails.gender}
- Comorbidities: ${patientDetails.comorbidities}
- Admission Date: ${new Date(patientDetails.admissionDate).toLocaleDateString()}

Patient Timeline Events:
${timelineDescription}

Task:
${promptInstruction}

Write a high-quality clinical narrative summary. Write in professional medical terminology. Do NOT mention LLMs, instructions, or prompts. Return ONLY the clinical narrative text.`
                }
              ]
            }
          ]
        }),
      }
    );

    if (!response.ok) {
      throw new Error(`Gemini API returned status ${response.status}`);
    }

    const data = await response.json();
    return data.candidates?.[0]?.content?.parts?.[0]?.text || 'Narrative generation failed.';
  } catch (error) {
    console.error('Error generating narrative via Gemini, falling back:', error);
    return mockExtractClinicalEvent(timelineEvents[timelineEvents.length - 1]?.eventData || '{}').summarySnippet;
  }
}

File: backend/app/utils/nlp.py
import json
import urllib.request
import urllib.error
from datetime import datetime

# Local heuristic parser for offline/no-key clinical event extraction
def mock_extract_clinical_event(text: str) -> dict:
    normalized = text.lower()
    
    # 1. Medications
    if any(k in normalized for k in ['start', 'give', 'administer', 'hold', 'stop', 'discontinue', 'resume']):
        status = 'started'
        if 'hold' in normalized or 'held' in normalized:
            status = 'held'
        elif 'stop' in normalized or 'discontinue' in normalized:
            status = 'stopped'
        elif 'resume' in normalized:
            status = 'started'
            
        drugs = ['meropenem', 'eliquis', 'solu-medrol', 'methylprednisolone', 'pantocid', 'lasix', 'levosalbutamol', 'budesonide', 'heparin', 'linezolid', 'norad', 'piptaz', 'vasopressin']
        found_drug = 'Unknown Medication'
        for d in drugs:
            if d in normalized:
                found_drug = d.capitalize()
                if d == 'solu-medrol':
                    found_drug = 'Solu-Medrol'
                break
                
        dose = 'standard'
        route = 'oral'
        freq = 'daily'
        
        if 'iv' in normalized: route = 'IV'
        if 'sc' in normalized or 'subcutaneous' in normalized: route = 'SC'
        if 'neb' in normalized: route = 'NEB'
        
        if '1g' in normalized or '1 g' in normalized: dose = '1g'
        elif '500mg' in normalized or '500 mg' in normalized: dose = '500mg'
        elif '5mg' in normalized or '5 mg' in normalized: dose = '5mg'
        elif '40mg' in normalized or '40 mg' in normalized: dose = '40mg'
        
        if 'tds' in normalized or 'three times' in normalized: freq = 'TDS'
        elif 'bd' in normalized or 'twice' in normalized: freq = 'BD'
        elif 'od' in normalized or 'once' in normalized: freq = 'OD'
        
        return {
            "eventType": "MEDICATION",
            "extractedEntities": {
                "drugName": found_drug,
                "dose": dose,
                "route": route,
                "freq": freq,
                "status": status,
                "reason": "active bleeding / safety hold" if status == "held" else "clinical management"
            },
            "confidence": 0.9,
            "flags": ["medication_hold"] if status == "held" else [],
            "summarySnippet": f"Medication {found_drug} {status} ({dose} {route} {freq})"
        }
        
    # 2. Investigations (Labs/Vitals)
    if any(k in normalized for k in ['hb', 'hemoglobin', 'creatinine', 'crp', 'wbc', 'platelet', 'abg', 'oxygen']):
        lab_name = 'Unknown Lab'
        value = 'normal'
        unit = ''
        status = 'Normal'
        flags = []
        
        if 'hb' in normalized or 'hemoglobin' in normalized:
            lab_name = 'Hemoglobin (Hb)'
            unit = 'g/dL'
            value = '7.5'  # Default mock value
            # Extract simple numeric value
            for word in normalized.split():
                try:
                    val = float(word.replace(',', ''))
                    if 4.0 < val < 18.0:
                        value = str(val)
                        break
                except ValueError:
                    pass
            if float(value) < 9.0:
                status = 'Critical Low'
                flags.append('critical_value')
        elif 'creatinine' in normalized:
            lab_name = 'Creatinine'
            unit = 'mg/dL'
            value = '2.4'  # Default mock
            for word in normalized.split():
                try:
                    val = float(word.replace(',', ''))
                    if 0.2 < val < 12.0:
                        value = str(val)
                        break
                except ValueError:
                    pass
            if float(value) > 1.5:
                status = 'High'
                flags.append('high_value')
        elif 'crp' in normalized:
            lab_name = 'CRP'
            unit = 'mg/L'
            value = '45'
            if 'crp' in normalized:
                status = 'Elevated'
                
        return {
            "eventType": "INVESTIGATION",
            "extractedEntities": {
                "labName": lab_name,
                "value": value,
                "unit": unit,
                "status": status,
                "notes": f"Seeded value."
            },
            "confidence": 0.95,
            "flags": flags,
            "summarySnippet": f"{lab_name} level: {value} {unit} ({status})"
        }

    # 3. Procedures
    if any(k in normalized for k in ['intubat', 'extubat', 'cath', 'dialysis', 'bronchoscopy']):
        proc_name = 'Procedure'
        if 'intubat' in normalized: proc_name = 'Endotracheal Intubation'
        elif 'extubat' in normalized: proc_name = 'Extubation'
        elif 'dialysis' in normalized or 'hemodialysis' in normalized: proc_name = 'Hemodialysis'
        elif 'bronchoscopy' in normalized: proc_name = 'Flexible Bronchoscopy'
        elif 'cath' in normalized: proc_name = 'Central Venous Catheterization'
        
        return {
            "eventType": "PROCEDURE",
            "extractedEntities": {
                "procedureName": proc_name,
                "status": "Completed",
                "notes": text
            },
            "confidence": 0.9,
            "flags": [],
            "summarySnippet": f"Procedure performed: {proc_name}"
        }

    # 4. Consultations
    if any(k in normalized for k in ['consult', 'referral', 'specialist', 'nephro', 'cardio', 'neuro']):
        spec = 'General Consultation'
        if 'nephro' in normalized: spec = 'Nephrology Consult'
        elif 'cardio' in normalized: spec = 'Cardiology Consult'
        elif 'neuro' in normalized: spec = 'Neurology Consult'
        
        return {
            "eventType": "CONSULTATION",
            "extractedEntities": {
                "specialty": spec,
                "notes": text
            },
            "confidence": 0.85,
            "flags": [],
            "summarySnippet": f"Specialist Consultation: {spec}"
        }

    # 5. Diagnosis
    if any(k in normalized for k in ['diagnos', 'admit', 'present with', 'pneumonia', 'bleeding', 'hemorrhage', 'copd', 'asthma']):
        diag = 'Clinical Condition'
        if 'pneumonia' in normalized: diag = 'Pneumonia'
        elif 'hemorrhage' in normalized or 'bleeding' in normalized: diag = 'Diffuse Alveolar Hemorrhage'
        elif 'ckd' in normalized or 'kidney' in normalized: diag = 'Chronic Kidney Disease'
        
        return {
            "eventType": "DIAGNOSIS",
            "extractedEntities": {
                "diagnosisName": diag,
                "status": "Active",
                "notes": text
            },
            "confidence": 0.85,
            "flags": [],
            "summarySnippet": f"Diagnosis recorded: {diag}"
        }

    return {
        "eventType": "NOTE",
        "extractedEntities": {
            "content": text
        },
        "confidence": 0.8,
        "flags": [],
        "summarySnippet": "Clinical progress update logged."
    }

# Vision model analysis simulator for CT/X-Ray image uploads
def mock_vision_image_extraction(filename: str) -> dict:
    fname = filename.lower()
    findings = "Bilateral chest consolidation, ground-glass opacities, resolving pulmonary edema."
    
    if 'xray' in fname or 'x-ray' in fname:
        findings = "Bilateral ground-glass opacities in lower zones. No pneumothorax. Heart size normal."
    elif 'ct' in fname:
        findings = "Chest CT confirms diffuse alveolar infiltrates and patchy consolidation bilaterally. Consistent with severe Diffuse Alveolar Hemorrhage (DAH)."
        
    return {
        "eventType": "INVESTIGATION",
        "extractedEntities": {
            "labName": "Radiological Imaging (Vision scan)",
            "value": findings,
            "unit": "Scan Findings",
            "status": "Abnormal",
            "notes": f"Scanned file metadata: {filename}"
        },
        "confidence": 0.9,
        "flags": ["abnormal_imaging"],
        "summarySnippet": f"Radiological Scan parsed: {findings[:50]}..."
    }

async def extract_clinical_event(text: str, api_key: str = None) -> dict:
    if not api_key or api_key in ("", "null", "undefined"):
        import os
        api_key = os.environ.get("GEMINI_API_KEY")
        
    if not api_key:
        return mock_extract_clinical_event(text)
        
    url = f"https://generativelanguage.googleapis.com/v1beta/models/gemini-2.5-flash:generateContent?key={api_key}"
    prompt = f"""You are an expert clinical documentation AI. Parse the following clinical text and extract structured medical event data in JSON format.
                  
Your output MUST be a valid JSON object matching this schema:
{{
  "eventType": "DIAGNOSIS" | "MEDICATION" | "PROCEDURE" | "INVESTIGATION" | "CONSULTATION" | "NOTE",
  "extractedEntities": {{
    // For DIAGNOSIS:
    "diagnosisName": string,
    "status": string ("Active", "Resolved", "Suspected"),
    // For MEDICATION:
    "drugName": string,
    "dose": string,
    "route": string,
    "freq": string,
    "status": string ("started", "held", "stopped", "continued"),
    "reason": string,
    // For PROCEDURE:
    "procedureName": string,
    "status": string,
    // For INVESTIGATION:
    "labName": string,
    "value": string,
    "unit": string,
    "status": string ("Normal", "High", "Low", "Critical Low"),
    // For CONSULTATION:
    "specialty": string,
    // For General NOTE:
    "content": string,
    "notes": string
  }},
  "confidence": number (0.0 to 1.0),
  "flags": string[] (e.g. "critical_value", "medication_hold", "conflict"),
  "summarySnippet": string (A one-line clinical summary sentence, e.g. "Eliquis held due to active pulmonary bleeding.")
}}

Return ONLY raw JSON. Do not include markdown code block formatting.

Clinical text to parse:
"{text}"
"""
    data = {
        "contents": [{"parts": [{"text": prompt}]}],
        "generationConfig": {"responseMimeType": "application/json"}
    }
    
    req = urllib.request.Request(
        url,
        data=json.dumps(data).encode('utf-8'),
        headers={'Content-Type': 'application/json'},
        method='POST'
    )
    
    try:
        with urllib.request.urlopen(req, timeout=10) as response:
            res_data = json.loads(response.read().decode('utf-8'))
            result_text = res_data['candidates'][0]['content']['parts'][0]['text']
            
            # Clean possible markdown block markers
            cleaned = result_text.strip()
            if cleaned.startswith("```json"):
                cleaned = cleaned[7:]
            if cleaned.endswith("```"):
                cleaned = cleaned[:-3]
                
            return json.loads(cleaned.strip())
    except Exception as e:
        print(f"Error in Gemini extract API, fallback to mock: {e}")
        return mock_extract_clinical_event(text)

async def generate_narrative(patient_details: dict, events: list, type_str: str, api_key: str = None) -> str:
    # Format timeline context
    timeline_desc = []
    for e in events:
        date_str = e.timestamp.strftime("%d %b %Y")
        try:
            data = json.loads(e.event_data)
        except Exception:
            data = {}
        timeline_desc.append(f"[{date_str}] ({e.event_type}) by {e.author_name}: {e.provenance} - {data.get('noteText') or json.dumps(data)}")
    timeline_text = "\n".join(timeline_desc)

    if not api_key or api_key in ("", "null", "undefined"):
        import os
        api_key = os.environ.get("GEMINI_API_KEY")

    if not api_key:
        # Static local template compiler
        meds = [e for e in events if e.event_type == 'MEDICATION']
        labs = [e for e in events if e.event_type == 'INVESTIGATION']
        procs = [e for e in events if e.event_type == 'PROCEDURE']
        
        if type_str == 'medicationJourney':
            med_text = "Active Medications Tracker:\n"
            for idx, m in enumerate(meds):
                try:
                    data = json.loads(m.event_data)
                except Exception:
                    data = {}
                med_text += f"{idx + 1}. {data.get('drugName')} {data.get('dose')} {data.get('route', '')} {data.get('freq', '')} - {data.get('status', '').upper()} ({data.get('reason', 'Management')})\n"
            return med_text if meds else "No active medications."
            
        elif type_str == 'investigationJourney':
            lab_text = "Laboratory Trends:\n"
            for l in labs:
                try:
                    data = json.loads(l.event_data)
                except Exception:
                    data = {}
                lab_text += f"- {l.timestamp.strftime('%Y-%m-%d')}: {data.get('labName')} {data.get('value')} {data.get('unit', '')} ({data.get('status')})\n"
            return lab_text if labs else "No lab investigations logged."
            
        elif type_str == 'procedureJourney':
            proc_text = "Procedures Ledger:\n"
            for p in procs:
                try:
                    data = json.loads(p.event_data)
                except Exception:
                    data = {}
                proc_text += f"- {p.timestamp.strftime('%Y-%m-%d')}: {data.get('procedureName')} - {data.get('status')}\n"
            return proc_text if procs else "No procedures performed."
            
        # Default Course in Hospital
        return f"Patient {patient_details['name']} ({patient_details['age']}yo {patient_details['gender']}), admitted on {patient_details['admission_date'].strftime('%Y-%m-%d')} with comorbidities: {patient_details['comorbidities']}.\n\n" + \
               "Hospital Course Timeline Notes:\n" + \
               "\n".join([f"- {e.timestamp.strftime('%b %d')}: {json.loads(e.event_data).get('summarySnippet') or e.provenance}" for e in events])

    url = f"https://generativelanguage.googleapis.com/v1beta/models/gemini-2.5-flash:generateContent?key={api_key}"
    
    prompt_instruction = ""
    if type_str == 'courseInHospital':
        prompt_instruction = "Generate a detailed, continuous, chronological narrative of the patient's hospital course in formal medical prose. Summarize diagnoses, clinical deterioration, stabilization, medication logic, laboratory results, and investigations. Include specialist consultations. Format as a cohesive paragraph."
    elif type_str == 'medicationJourney':
        prompt_instruction = "Generate a summarized Medication Journey. List and explain started, held, and stopped medications with their respective clinical reasons based on the timeline."
    elif type_str == 'investigationJourney':
        prompt_instruction = "Generate a summary of laboratory and diagnostic investigation trends. Focus on key biometrics (e.g., Creatinine, Hemoglobin, CRP) and mention values chronologically."
    else:
        prompt_instruction = "Generate a chronological list of procedures performed on the patient, detailing their status and indications."

    prompt = f"""You are an elite pulmonology and critical care clinical documentation assistant.
                  
Patient Details:
- Name: {patient_details['name']}
- Age: {patient_details['age']}
- Gender: {patient_details['gender']}
- Comorbidities: {patient_details['comorbidities']}
- Admission Date: {patient_details['admission_date'].strftime('%Y-%m-%d')}

Patient Timeline Events:
{timeline_text}

Task:
{prompt_instruction}

Write a high-quality clinical narrative summary in professional medical terminology. Do NOT mention LLMs, instructions, or prompts. Return ONLY the clinical narrative text.
"""
    
    data = {
        "contents": [{"parts": [{"text": prompt}]}]
    }
    
    req = urllib.request.Request(
        url,
        data=json.dumps(data).encode('utf-8'),
        headers={'Content-Type': 'application/json'},
        method='POST'
    )
    
    try:
        with urllib.request.urlopen(req, timeout=15) as response:
            res_data = json.loads(response.read().decode('utf-8'))
            return res_data['candidates'][0]['content']['parts'][0]['text']
    except Exception as e:
        print(f"Error generating narrative: {e}")
        return "Narrative generation failed."
